File: src/main/java/core/procedures/generic/Next.java
package core.procedures.generic;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Cons;
import core.scm.Vector;
import core.utils.Utils;

import java.util.*;

public class Next extends AFn {

  public Next() {
    super(new FnArgsBuilder().min(1).max(1).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "next";
  }

  @Override
  public Object apply1(Object arg) {
    if (!Utils.isSeqable(arg)) {
      throw new IllegalArgumentException("don't know how to create Sequence from " + arg.getClass());
    }
    if (arg instanceof List) {
      List list = (List) arg;
      return list.isEmpty() ? null : list.subList(1, list.size());
    } else if (arg instanceof Set) {
      return next((Set) arg);
    } else if (arg instanceof Map) {
      return next(((Map) arg).entrySet());
    } else if (arg instanceof Map.Entry) {
      return Cons.list(((Map.Entry) arg).getValue());
    } else if (arg instanceof Vector) {
      Vector vec = (Vector) arg;
      return vec.size() == 0 ? null : new Vector(Arrays.copyOfRange(vec.getArray(), 1, vec.size()));
    } else if (arg instanceof CharSequence) {
      CharSequence cs = (CharSequence) arg;
      return cs.length() == 0 ? null : cs.subSequence(1, cs.length());
    }
    throw new WrongTypeException("next", "List or Vector or Set or String or Map", arg);
  }

  private Object next(Set set) {
    if (set.isEmpty()) {
      return null;
    }
    Set<Object> next = new HashSet<>();
    Iterator iter = set.iterator();
    iter.next();
    while (iter.hasNext()) {
      next.add(iter.next());
    }
    return next;
  }
}

File: src/main/java/core/procedures/generic/Next.kt
package core.procedures.generic

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.scm.Cons
import core.scm.MutableVector
import core.scm.Vector
import core.utils.Utils

open class Next : AFn<Any?, Any?>(name = "next", isPure = true, minArgs = 1, maxArgs = 1) {

    override operator fun invoke(arg: Any?): Any? = when (arg) {
        null                  -> null
        !Utils.isSeqable(arg) -> throw IllegalArgumentException("don't know how to create Sequence from ${arg.javaClass}")
        is List<*>            -> if (arg.isEmpty()) null else arg.subList(1, arg.size)
        is Set<*>             -> next(arg)
        is Map<*, *>          -> next(arg.entries)
        is Map.Entry<*, *>    -> Cons.list(arg.value!!)
        is Vector             -> if (arg.size == 0) null else MutableVector(arg)
        is CharSequence       -> if (arg.length == 0) null else arg.subSequence(1, arg.length)
        is ByteArray          -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is BooleanArray       -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is CharArray          -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is ShortArray         -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is IntArray           -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is LongArray          -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is DoubleArray        -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is FloatArray         -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        is Array<*>           -> if (arg.size == 0) null else arg.copyOfRange(1, arg.size)
        else                  -> throw WrongTypeException("next", "List or Vector or Set or String or Map", arg)
    }

    private fun next(col: Collection<*>) = when {
        col.isEmpty() -> null
        else -> col.asSequence().drop(1).toSet()
    }
}
